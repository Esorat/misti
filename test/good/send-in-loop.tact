message Msg {
    a: Int;
}

contract T {
    fun tWhile(){
        let x: Int = 10;
while (x > 0) {
  // Bad: x is not changed due looping
  send(SendParameters{ to: sender(), value: x });
}
    }
}

contract Z {
    fun zWhile(){
        let x: Int = 10;
        while (x > 0) {
            // OK: Should not trigger!
            send(SendParameters{ to: sender(), value: x });
            x = x - 1; // Decrement inside the loop
        }
    }
}


contract SendInLoopExamples {
    owner: Address;
    a: Int;

    init(owner: Address) {
        self.owner = owner;
        self.a = 0;
    }

    // Example 1: Send function inside a while loop
    fun exampleWhileLoop(limit: Int) {
        let i = 0;
        while (i < limit) {
            i += 1;
            send(SendParameters{
                to: self.owner,
                value: 0,
                bounce: false,
                mode: SendRemainingValue,
                body: Msg{ a: i }.toCell()
            });
        }
    }

    // Example 2: Send function inside a repeat loop
    fun exampleRepeatLoop() {
        let i = 0;
        repeat (5) {
            i += 1;
            send(SendParameters{
                to: self.owner,
                value: 0,
                bounce: false,
                mode: SendRemainingValue,
                body: Msg{ a: i }.toCell()
            });
        }
    }

    // Example 3: Send function inside nested loops
    fun exampleNestedLoops() {
        let i = 0;
        while (i < 3) {
            i += 1;
            let j = 0;
            while (j < 2) {
                j += 1;
                send(SendParameters{
                    to: self.owner,
                    value: 0,
                    bounce: false,
                    mode: SendRemainingValue,
                    body: Msg{ a: i * 10 + j }.toCell()
                });
            }
        }
    }

    // Example 4: Send function inside a loop within an if condition
    fun exampleLoopInIfCondition(flag: Bool) {
        if (flag) {
            let i = 0;
            while (i < 4) {
                i += 1;
                send(SendParameters{
                    to: self.owner,
                    value: 0,
                    bounce: false,
                    mode: SendRemainingValue,
                    body: Msg{ a: i }.toCell()
                });
            }
        }
    }

    // Example 5: Send function inside a loop with adjusted condition
    fun exampleInfiniteLoop() {
        let i = 0;
        while (i < 10) { // Adjusted condition to prevent infinite loop
            i += 1;
            send(SendParameters{
                to: self.owner,
                value: 0,
                bounce: false,
                mode: SendRemainingValue,
                body: Msg{ a: i }.toCell()
            });
        }
    }

    // Example 6: Send function inside a loop within a try-catch block
    fun exampleTryCatchLoop() {
        try {
            let i = 0;
            while (i < 5) {
                i += 1;
                send(SendParameters{
                    to: self.owner,
                    value: 0,
                    bounce: false,
                    mode: SendRemainingValue,
                    body: Msg{ a: i }.toCell()
                });
            }
        } catch (e) {
            // Handle exception
        }
    }


    // Example 8: Send function outside of loops (should not trigger warning)
    fun exampleNoLoop() {
        send(SendParameters{
            to: self.owner,
            value: 0,
            bounce: false,
            mode: SendRemainingValue,
            body: Msg{ a: 0 }.toCell()
        });
    }

}

contract Test {
    fun tWhile() { 
        let i: Int = 0;
        while (i < 10) {
            send(SendParameters{
                to: sender(),
                value: i,
            });
            i = i + 1;
        }
    }

    fun tRepeat() { 
        repeat (10) {
            send(SendParameters{
                to: sender(),
                value: 0,
            });
        }
    }

    fun tUntil() {
        let i: Int = 0;
        do {
            send(SendParameters{
                to: sender(),
                value: i,
            });
            i = i + 1;
        } until (i == 10);
    }

    fun tForeach() {
        let m: map<Int, Int> = emptyMap();
        foreach (k, v in m) {
            send(SendParameters{
                to: sender(),
                value: v,
            });
        }
    }

    fun tNested() {
        let i: Int = 0;
        while (i < 10) {
            repeat (10) {
                send(SendParameters{
                    to: sender(),
                    value: i,
                });
            }
            i = i + 1;
        }
    }
}