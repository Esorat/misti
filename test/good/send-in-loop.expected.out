[HIGH] SendInLoop: Send function 'send' called inside a loop
test/good/send-in-loop.tact:5:13:
  4 |         while (i < 10) {
> 5 |             send(SendParameters{
                  ^
  6 |                 to: sender(),
Help: Consider refactoring to avoid calling send functions inside loops.
See: https://nowarp.io/tools/misti/docs/detectors/SendInLoop

[HIGH] SendInLoop: Send function 'send' called inside a loop
test/good/send-in-loop.tact:15:13:
  14 |         repeat (10) {
> 15 |             send(SendParameters{
                   ^
  16 |                 to: sender(),
Help: Consider refactoring to avoid calling send functions inside loops.
See: https://nowarp.io/tools/misti/docs/detectors/SendInLoop

[HIGH] SendInLoop: Send function 'send' called inside a loop
test/good/send-in-loop.tact:25:13:
  24 |         do {
> 25 |             send(SendParameters{
                   ^
  26 |                 to: sender(),
Help: Consider refactoring to avoid calling send functions inside loops.
See: https://nowarp.io/tools/misti/docs/detectors/SendInLoop

[HIGH] SendInLoop: Send function 'send' called inside a loop
test/good/send-in-loop.tact:36:13:
  35 |         foreach (k, v in m) {
> 36 |             send(SendParameters{
                   ^
  37 |                 to: sender(),
Help: Consider refactoring to avoid calling send functions inside loops.
See: https://nowarp.io/tools/misti/docs/detectors/SendInLoop

[HIGH] SendInLoop: Send function 'send' called inside a loop
test/good/send-in-loop.tact:47:17:
  46 |             repeat (10) {
> 47 |                 send(SendParameters{
                       ^
  48 |                     to: sender(),
Help: Consider refactoring to avoid calling send functions inside loops.
See: https://nowarp.io/tools/misti/docs/detectors/SendInLoop

[INFO] PreferAugmentedAssign: Prefer augmented assignment: i += 1
test/good/send-in-loop.tact:9:13:
   8 |             });
>  9 |             i = i + 1;
                   ^
  10 |         }
Help: Consider using augmented assignment instead: i += 1
See: https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign

[INFO] PreferAugmentedAssign: Prefer augmented assignment: i += 1
test/good/send-in-loop.tact:29:13:
  28 |             });
> 29 |             i = i + 1;
                   ^
  30 |         } until (i == 10);
Help: Consider using augmented assignment instead: i += 1
See: https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign

[INFO] PreferAugmentedAssign: Prefer augmented assignment: i += 1
test/good/send-in-loop.tact:52:13:
  51 |             }
> 52 |             i = i + 1;
                   ^
  53 |         }
Help: Consider using augmented assignment instead: i += 1
See: https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign